### PR TITLE
Fix and de-risk UserVmManagerImpl external-DHCP startup scan

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

UserVmManagerImpl.start() synchronously ran loadVmDetailsInMapForExternalDhcpIp() on the management-server startup thread. Because the CloudStack lifecycle runs every bean's start() sequentially on main before the agent listener (port 8250) is bound, this scan sat directly on the path to agent readiness.

Actual behaviour: on large, DB-contended clusters the scan blocked the agent listener bind for many minutes: 0.8 - 36.9 min observed across some prod environment and repeatedly crashed with a NullPointerException in NetworkModelImpl.isSharedNetworkWithoutServices when a network was removed mid-scan (10 occurrences on 7 hosts / 3 clusters).

Expected behaviour: the management server becomes agent-ready without this scan gating startup, and a concurrently-removed network never crashes it.

Changes:

NetworkModelImpl.isSharedNetworkWithoutServices: guard against a null network from findById (fixes the NPE for all callers).
UserVmManagerImpl.start(): seed the external-DHCP map asynchronously so it no longer gates agent readiness (it is only a warm-start cache for VmIpFetchTask).
loadVmDetailsInMapForExternalDhcpIp: replace the per-NIC findById N+1 with a batched listByIds, memoize the per-offering "without services" check, and add a Profiler timing summary line.
Add unit tests for the null-guard, the batched/filtered seeding, and the async submission.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] Build/CI
- [ ] Test (unit or integration test code)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

File: engine/schema/src/main/java/com/cloud/vm/dao/VMInstanceDaoImpl.java [hunk 1]
```diff
@@ -1261,4 +1261,17 @@ public int getVmCountByOfferingNotInDomain(Long serviceOfferingId, List<Long> do
         List<Integer> count = customSearch(sc, null);
         return count.get(0);
     }
+
+    @Override
+    public List<VMInstanceVO> listByIds(List<Long> ids) {
+        if (ids == null || ids.isEmpty()) {
```
Comment (weizhouapache) [inline engine/schema/src/main/java/com/cloud/vm/dao/VMInstanceDaoImpl.java]: CollectionUtils ?

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing - tested with 5 shared networks (with and without DHCP service) and 20 VMs on different networks, MS/networks restarts.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -2452,33 +2453,53 @@ public String getName() {
...
+            int networksScanned = 0;
+            int nicsAdded = 0;
+
+            for (NetworkVO network : networks) {
+                boolean withoutServices = offeringWithoutServices.computeIfAbsent(network.getNetworkOfferingId(),
+                        offeringId -> _networkModel.listNetworkOfferingServices(offeringId).isEmpty());
+                if (!withoutServices) {
+                    continue;
+                }
+                networksScanned++;
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2478]: seems valid concern

File: engine/schema/src/main/java/com/cloud/vm/dao/VMInstanceDaoImpl.java [hunk 3]
```diff
@@ -1261,4 +1262,17 @@ public int getVmCountByOfferingNotInDomain(Long serviceOfferingId, List<Long> do
         List<Integer> count = customSearch(sc, null);
         return count.get(0);
     }
+
+    @Override
+    public List<VMInstanceVO> listByIds(List<Long> ids) {
+        if (CollectionUtils.isEmpty(ids)) {
+            return Collections.emptyList();
+        }
+        SearchBuilder<VMInstanceVO> sb = createSearchBuilder();
```
Comment (weizhouapache) [inline engine/schema/src/main/java/com/cloud/vm/dao/VMInstanceDaoImpl.java:1271]: create a static SearchBuilder ?

**Comment by Pearl1594** [comment]
@blueorangutan package

File: engine/schema/src/main/java/com/cloud/vm/dao/VMInstanceDaoImpl.java (hunk 3 above)
Comment (Pearl1594) [reply engine/schema/src/main/java/com/cloud/vm/dao/VMInstanceDaoImpl.java:1271]: Im not sure what you mean @weizhouapache

**Review by vladimirpetrov — APPROVED** [review]
Latest change LGTM based on manual testing, tested with L2 networks with and without ConfigDrive, looks fine.
